app.py

```diff
@@ -83,7 +87,45 @@
 (NBEST, CONSTRAINTS, DECTREE, REATTACH, RELABEL, REPARSE, EDITDIST, TIME
 		) = range(8)
 # e.g., "NN-SB/Nom" => ('NN', '-SB', '/Nom')
+
+PUNCT_ESCAPING = [{"istring" : "(", "ptree_label": "LRB-p", "ptree_token": "-LRB-", "ctree_punct": "("},
+				  {"istring" : ")", "ptree_label": "RRB-p", "ptree_token": "-RRB-", "ctree_punct": ")"},
+				  {"istring" : "-", "ptree_label": "HYPH-p", "ptree_token": "-", "ctree_punct": "-"}]
+
 LABELRE = re.compile(r'^([^-/\s]+)(-[^/\s]+)?(/\S+)?$')
+PUNCTRE = re.compile(r'^(\W+)$')
+PUNCTRE_LABEL = re.compile(r'^(\W+)-p$')
+INITIAL_PUNCT_LABEL = {'LRB-p', '[-p', '{-p'}
+AMBIG_SYM = {'$', '#', '@', '&', '-'}
+
+def is_possible_punct_token(token):
+	return re.match(PUNCTRE, token) or token in [e['ptree_token'] for e in PUNCT_ESCAPING]
+
+def is_punct_label(label):
+	return re.match(PUNCTRE_LABEL, label) or label in [e['ptree_label'] for e in PUNCT_ESCAPING]
+
+def sent_escape(sent):
+	"""Replace special characters in a sentence. (First splits the sentence into tokens.)
+	If a token is an 'istring' property of a PUNCT_ESCAPING element, replace it with the 'ptree_token' property."""
+	senttok = sent.split()
+	for i, token in enumerate(senttok):
+		for e in PUNCT_ESCAPING:
+			if token == e['istring']:
+				senttok[i] = e['ptree_token']
+				break
+	return " ".join(senttok)
+
+def senttok_escape(senttok):
+	"""Replace special characters in a tokenized sentence.
+	If a token is an 'istring' property of a PUNCT_ESCAPING element, replace it with the 'ptree_token' property."""
+	senttok = [t for t in senttok]
```

Comment:
I guess the point is to create a copy of the list—slightly more idiomatic is to use the copy constructor: `list(senttok)`